HIP/hip-0000-atomic-association-contracts.md

```diff
@@ -0,0 +1,104 @@
+---
+hip: 0000
+title: Waived association requirement for atomic transfers for smart contracts
+author: Matthew DeLorenzo <mdelore.ufl@gmail.com>, Vae Vecturne <vea.vecturne@gmail.com>
+type: Standards Track
+category: Service
+needs-council-approval: Yes
+status: Draft
+created: 2023-04-06
+discussions-to: https://github.com/hashgraph/hedera-improvement-proposal/discussions/696
+---
+
+## Abstract
+
+Smart contracts are required to be associated with all tokens they custody. This makes sense when the contract has a long lived balance > 0; Hedera state is expanded and must be paid for in ongoing maintenance and storage costs to the network. However, there are many use cases that call for smart contracts atomically taking custody of a token, often to transfer to another entity within the same contract call. In this case, association is still required but Hedera state is not expanded.
+
+We propose that association not be required for Hedera entities that only custody a token during an atomic transaction.
+
+## Motivation
+
+Currently, a smart contract attempting to custody an unassociated token will be immediately rejected by the network. This presents a problem in some decentralized finance (DeFi) applications that use smart contracts to atomically custody tokens. Decentraliztion may be achieved by allowing contracts to remain token agnostic, but also prevent spamming attacks by allowing any user to associate tokens to a DeFi smart contract. 
+
+## Rationale
+
+Periphery contracts for decentralized exchanges (DEXes), such as Uniswap V2 and V3, are designed to atomically custody tokens. For DEXes on Hedera, gas-inefficient workarounds are used to associate and dissociate tokens to periphery smart contracts, paid by end-users.
+
+## Specification
+
+Hedera Smart Contract Service optimistically allows custody of any token for smart contracts, and checks at the end of a contract call that tokens are associated to entities whose balance is changing from zero to nonzero. 
+
+### Successful
+
+The response code for token transfer of a token not associated to a smart contract, whose balance is equal to zero at the end of the transaction, is SUCCESS = 22.
+
+### Revert
+
+The response code for token transfer of a token not associated to a smart contract, whose balance is nonzero at the end of the transaction, is TOKEN_NOT_ASSOCIATED_TO_ACCOUNT = 184, or a new enum TOKEN_NOT_ASSOCIATED_TO_CONTRACT.
+
+## User stories
+
+As a solidity developer, I do not want to associate tokens to a smart contract that custodies tokens but is not designed to have token balances.
+
+## Backwards Compatibility
+
+This would be backward compatible.
+
+## Security Implications
+
+Increased complexity of Hedera Smart Contract Service may lead to vulnerabilities.
+
+## How to Teach This
+
+N/A
+
+## Reference Implementation
+#### Example :
```

Comment:
In the code examples below, I'd recommend:

- explicitly define the input parameters (`token` and `recipient`)
  - e.g. can `recipient` only be an EOA, or could it be any address, including SCs?
  - e.g. can `token` only be an HTS token, or could it be any token, including ones authored in Solidity?
  - `@param treasury` should be `@param recipient` in the function-level comments
- in lieu of the comment `// assumes msg.sender approved an allowance to address(this)`, include the code for the approval, as it is more explicit
- In `function1`, the return value of the 2nd invocation of  `HederaTokenService.transferToken` is saved in `respCode`, whereas in `function2` this does not happen
  - I assume that you wanted to do so in both, and assert that in `function1` it was `SUCCESS`, and in `function2` it should be `TOKEN_NOT_ASSOCIATED_TO_ACCOUNT` or your newly proposed `TOKEN_NOT_ASSOCIATED_TO_CONTRACT`
  - If so, add that assertion or a comment; plus consider adding `function3` which results in the newly proposed response code.